### PR TITLE
WebXRManager: Add AR plane detection.

### PR DESCRIPTION
**Pull request by richardanaya**
Related #24792.

**Description**

Added support for Meta Quest Pro and Quest 2 support for plane detection API along with an example.

image

### CODE REVIEW

**Comment by Mugen87** [comment]
For testing: https://raw.githack.com/richardanaya/three.js/dev/examples/webxr_ar_plane_detection.html

**Comment by Mugen87** [comment]
Please add the new example to https://github.com/mrdoob/three.js/blob/dev/examples/files.json so it appears at the side menu.

Also create a thumbnail via `npm run make-screenshot webxr_ar_plane_detection`.

**Comment by richardanaya** [comment]
@Mugen87 Added `files.json`  and screenshot :) 

I think there might be a problem with that link you sent to test since this example depends upon a compiled threejs.  I made a test branch here that has my change to WebXRManager compiled in.

https://raw.githack.com/richardanaya/three.js/test/examples/webxr_ar_plane_detection.html

**Comment by Mugen87** [comment]
Thanks for updating the link!

How do I test this with a Quest 2? I've only used the device for VR so far. When opening the example, I see the same like on my desktop (black screen and the "AR not supported" message).

**Comment by Mugen87** [comment]
Unlike the other AR example, I can't use the new one on a Pixel 4A with latest Chrome. Clicking on the "Start AR" button does nothing.

**Comment by richardanaya** [comment]
@Mugen87 right now there's a step you have to do for Quest:

1. Go to Settings -> Experimental -> Setup Room
2. Add some walls and doors or anything.
3. Restart device and you should see the option for AR.

If it works in the reference demo:
https://immersive-web.github.io/webxr-samples/proposals/plane-detection.html

it should work in:
https://raw.githack.com/richardanaya/three.js/test/examples/webxr_ar_plane_detection.html

**Comment by Mugen87** [comment]
Thanks for the instruction! Indeed it works as expected. 

/cc @cabanier

**Comment by cabanier** [comment]
> 1. Go to Settings -> Experimental -> Setup Room
> 2. Add some walls and doors or anything.
> 3. Restart device and you should see the option for AR when you go to try examples in Quest Browser

You don't need to restart the device after setting up walls and doors. They are pulled each time you start an immersive session.

**Comment by cabanier** [comment]
> Thanks for the instruction! Indeed it works as expected.

Does this mean that you get a plane event each frame?

**Comment by Mugen87** [comment]
>> Thanks for the instruction! Indeed it works as expected.

> Does this mean that you get a plane event each frame?

I've wanted to say that the rendering looks as expected. Meaning the configured walls are correctly displayed.

The new `detectedPlanes` event is fired each frame. Isn't that optimal?

**Comment by cabanier** [comment]
> > > Thanks for the instruction! Indeed it works as expected.
> 
> > Does this mean that you get a plane event each frame?
> 
> I've wanted to say that the rendering looks as expected. Meaning the configured walls are correctly displayed.
> 
> The new `detectedPlanes` event is fired each frame. Isn't that optimal?

It might be nice to get an event each time the planes change.
On Quest, this is only once. On Android this happens constantly.

**Comment by richardanaya** [comment]
@Mugen87 added some new events as per @rcabanier 's recommendation

**Comment by richardanaya** [comment]
Fixed a small issue related to removing planes and trying to conserve allocations.  Should be good to merge to my knowledge.

**Comment by richardanaya** [comment]
added `getKnownPlanes` method as a way to acquire known planes outside of event responses.

**Comment by richardanaya** [comment]
@Mugen87 I went through once more, and tried to make sure style in example and code was consistent. I think this is ready for a review.

File: src/renderers/webxr/WebXRManager.js [hunk 1]
```diff
@@ -600,6 +603,12 @@ class WebXRManager extends EventDispatcher {
 
 		};
 
+		this.getKnownPlanes = function () {
```
Comment (Mugen87) [inline src/renderers/webxr/WebXRManager.js:606]: For simplicity, I would rename `getKnownPlanes()` to `getPlanes()`. Also the internal variables `knownPlanes` and `knownPlaneLastChangedTimes` could be just `planes` and `planesLastChangedTimes`.

File: examples/webxr_ar_plane_detection.html [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+							minZ = Math.min( minZ, point.z );
+							maxZ = Math.max( maxZ, point.z );
+
+						} );
+
+						const width = maxX - minX;
+						const height = maxZ - minZ;
+
+						const boxMesh = new THREE.Mesh(
+							new THREE.BoxBufferGeometry( width, 0.01, height ),
```
Comment (Mugen87) [inline examples/webxr_ar_plane_detection.html:131]: Please do not use the `*BufferGeometry` aliases anymore. They are deprecated. Just use `BoxGeometry`.

**Comment by richardanaya** [comment]
@Mugen87 resolved requests with name changes and re-tested and updated test branch :)

**Comment by richardanaya** [comment]
Added documentation.

**Comment by cabanier** [comment]
LGTM!

**Comment by mrdoob** [comment]
@richardanaya Seems like you can have multiple rooms if you enter/exit XR.

Should the example delete the previous generated geometry when entering XR?

**Comment by mrdoob** [comment]
> @richardanaya Seems like you can have multiple rooms if you enter/exit XR.
>
> Should the example delete the previous generated geometry when entering XR?

Just took care of this in #26098

**Pull request merged**